server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java

```diff
@@ -337,12 +342,21 @@ void buildDiffAndSerializeStates() {
                     // publication to local node bypasses any serialization
                     continue;
                 }
+                TransportVersion version;
+                try {
+                    version = transportService.getConnection(node).getTransportVersion();
```

Comment:
ah, but we should keep hold of the connection we get here for later use, rather than calling `transportService.getConnection()` again further down.

Reply:
Ah, I see what you mean now. That is a more involved change, and at its core `getConnection` is just a concurrentmap access, so calling it twice is not toooooo bad. Do we need to cache it inside the context here?

Reply:
It's not the performance that worries me so much as the consistency. If we aren't connected to a node here but we _are_ connected by the time we call `sendClusterStateDiff` then we might trip the `assert bytes != null` therein. And if we use a different `Connection` later then in principle it could have a different transport version.

Reply:
Ok - added in [99339e186ee](https://github.com/elastic/elasticsearch/pull/93822/commits/99339e186ee07aff922f97d212895b7d23c54b4b)